# 01 숫자형

### 정수형

In [2]:
a=123
a=-178
a=0

### 실수형

In [3]:
a=1.2
a=-3.45

### 8진수

In [4]:
a=0o177
print(a)

127


### 16진수

In [5]:
a=0x8ff
b=0xabc
print(b)

2748


### 사칙연산

In [1]:
a=3
b=4
a+b

7

In [2]:
a-b

-1

In [3]:
a*b

12

In [4]:
a/b

0.75

### x의 y제곱을 나타내는 ** 연산자 (x=a, y=b)

In [10]:
a=3
b=4
a**b

81

### 나눗셈 후 나머지를 리턴하는 % 연산자
<img src="https://wikidocs.net/images/page/12/02_1_mod1.png" align="left"></img>

In [11]:
3%7

3

In [8]:
7%3

1

### 나눗셈 후 몫을 리턴하는 // 연산자
<img src="https://wikidocs.net/images/page/12/02_1_mod2.png" align="left"></img>

In [23]:
7/4

1.75

In [24]:
7//4

1

1.75에서 몫에 해당하는 정숫값 1만 리턴하는 것을 확인할 수 있다.

----

# 02 문자열 자료형

In [11]:
"hello world"

'hello world'

"a" >> 'a'

In [12]:
'hello world'

'hello world'

'a >> 'a'

# 문자열 안에 작은따옴표나 큰따옴표를 포함시키고 싶을 때

In [13]:
"Python's favorite food is perl"

"Python's favorite food is perl"

"a's" >> "a's"

In [18]:
'"Python is very easy." he says.'

'"Python is very easy." he says.'

' "abc" ' >> ' "abc" '

In [19]:
'Python\'s favorite food is perl'

"Python's favorite food is perl"

'a\'s' >> "a's"

In [2]:
"\"Python is very easy.\" he says."

'"Python is very easy." he says.'

"\"abc.\"defg" >> '"abc."defg'

### 줄바꿈

In [8]:
"Life is too short\nYou need python"

'Life is too short\nYou need python'

In [6]:
multiline = "Life is too short\nYou need python"

In [7]:
print(multiline)

Life is too short
You need python


In [9]:
multiline = '''
Life is too short
You need python
'''

In [10]:
multiline

'\nLife is too short\nYou need python\n'

In [11]:
print(multiline)


Life is too short
You need python



### 문자열 연산하기

In [12]:
head = "Python"
tail = " is fun!"
head + tail

'Python is fun!'

In [13]:
a='python'
a*5

'pythonpythonpythonpythonpython'

### 문자열 곱하기를 응용하기

### 문자열 길이 구하기

In [16]:
a="life is too short"
len(a)

17

### 문자열 인덱싱 (파이썬은 0부터 숫자를 센다)

In [17]:
a = "Life is too short, You need Python"
a[3]

'e'

In [18]:
a[12]

's'

### - 는 뒤에서 부터

In [19]:
a[-1]

'n'

In [20]:
a[-7]

' '

### 문자열 슬라이싱 (end 인덱스 +1 해야한다)

In [25]:
a = "Life is too short, You need Python"
a[0:3]

'Lif'

In [29]:
a[0:4]

'Life'

In [30]:
a[8:17]

'too short'

In [32]:
a[0:4]==a[0:5]

False

In [33]:
a[19:]

'You need Python'

In [34]:
a[:17]

'Life is too short'

In [36]:
a[:]

'Life is too short, You need Python'

In [38]:
a[19:-7]

'You need'

### 슬라이싱으로 문자열 나누기

In [40]:
a = "20230331Rainy"
date = a[:8]
weather = a[8:]
date, weather

('20230331', 'Rainy')

In [52]:
a = "20230331Rainy"
year = a[:4]
day = a[4:8]
weather = a[8:]
year

'2023'

In [53]:
day

'0331'

In [54]:
weather

'Rainy'

In [51]:
#문자열 수정
a='pithon'
a[1]=y
a

#당연히 오류가 발생한다. 문자열의 요솟값은 바꿀 수 있는 값이 아니기 때문이다
#(그래서 문자열을 ‘변경 불가능한(immutable) 자료형’이라고도 부른다).

NameError: name 'y' is not defined

In [44]:
a[:1]

'p'

In [45]:
a[2:]

'thon'

In [47]:
b= a[:1]+'y'+a[2:]
b

'python'

### 문자열 포매팅

In [56]:
#숫자바로대입
"I eat %d apples."%3

'I eat 3 apples.'

In [58]:
#문자열 바로대입
"I eat %s apples."%"five"

'I eat five apples.'

In [59]:
"I eat %d apples."%삼

NameError: name '삼' is not defined

In [64]:
#숫자값을 나타내는 변수로 대입
number=3
"I eat %d apples."%number

'I eat 3 apples.'

In [66]:
#2개 이상의 값 넣기
number = 10
day ='three'
"I ate %d apples. so I was sick for %s days." % (number, day)

'I ate 10 apples. so I was sick for three days.'

In [68]:
#d>>숫자 안됌, s>>문자로 인지가 됌
"I have %s apples" % 3

'I have 3 apples'

In [70]:
"오늘의 습도는%d%% 입니다." %70

'오늘의 습도는70% 입니다.'

### 포맷 코드와 숫자 함께 사용하기

In [71]:
#정렬과 공백
"%10s" % "hi"

'        hi'

In [72]:
"%-10sjane." % 'hi'

'hi        jane.'

### 소수점 표현하기

In [73]:
"%0.4f" % 3.42134234

'3.4213'

In [74]:
"%10.4f" % 3.42134234

'    3.4213'

### format 함수를 사용한 포매팅

In [77]:
"I eat {} apples".format(3)

'I eat 3 apples'

In [76]:
"I eat {} apples".format('five')

'I eat five apples'

In [78]:
number=10
"I eat {} apples".format(number)

'I eat 10 apples'

In [81]:
number = 10
day = "three"
"I ate {} apples. so I was sick for {} days.".format(number, day)

'I ate 10 apples. so I was sick for three days.'

In [82]:
"I ate {number} apples. so I was sick for {day} days.".format(number=10, day=3)

'I ate 10 apples. so I was sick for 3 days.'

In [83]:
 "I ate {0} apples. so I was sick for {day} days.".format(10, day=3)

'I ate 10 apples. so I was sick for 3 days.'

In [84]:
#소수점 표현하기
y = 3.42134234
"{0:0.4f}".format(y)

'3.4213'

### f 문자열 포매팅 (강추)

In [85]:
name = '홍길동'
age = 30
f'나의 이름은 {name}입니다. 나이는 {age}입니다.'

'나의 이름은 홍길동입니다. 나이는 30입니다.'

In [86]:
age = 30
f'나는 내년이면 {age + 1}살이 된다.'

'나는 내년이면 31살이 된다.'

In [87]:
d = {'name':'홍길동', 'age':30}
f'나의 이름은 {d["name"]}입니다. 나이는 {d["age"]}입니다.'

'나의 이름은 홍길동입니다. 나이는 30입니다.'

### 문자열 관련 함수들

In [88]:
#문자 개수 세기 - count
a = "hobby"
a.count('b')

2

In [89]:
#위치 알려 주기 1 - find
a = "Python is the best choice"
a.find('b')

14

In [90]:
#위치 알려 주기 2 - index 인덱스는 오류를 보여준다
a = "Life is too short"
a.index('t')

8

In [91]:
a.index('k')

ValueError: substring not found

In [92]:
#문자열 삽입 - join
",".join('abcd')

'a,b,c,d'

In [93]:
#소문자를 대문자로 바꾸기 - upper
a = "hi"
a.upper()

'HI'

In [95]:
#대문자를 소문자로 바꾸기 - lower
a="HI"
a.lower()

'hi'

In [96]:
#왼쪽 공백 지우기 - lstrip
a= " hi "
a.lstrip()

'hi '

In [97]:
#오른쪽 공백 지우기 - rstrip
a= " hi "
a.rstrip()

' hi'

In [98]:
#양쪽 공백 지우기 - strip
a= " hi "
a.strip()

'hi'

In [99]:
#문자열 바꾸기 - replace
a = "Life is too short"
a.replace('Life','Your leg')

'Your leg is too short'

In [100]:
a

'Life is too short'

In [101]:
#문자열 나누기 - split
a = "Life is too short"
a.split()

['Life', 'is', 'too', 'short']

In [102]:
b='a:b:c:d'
b.split(':')

['a', 'b', 'c', 'd']

# 변수

In [104]:
a=1
a, type(a)

(1, int)

In [106]:
b='python'
b, type(b)

('python', str)

In [108]:
c=[1,2,3]
c, type(c)

([1, 2, 3], list)

In [109]:
id(a)

140714777809704

In [117]:
#문자열을 복사
a=[1,2,3]
b=a

In [118]:
id(a), id(b)

(2753595138944, 2753595138944)

In [119]:
a is b

True

In [120]:
del b

In [121]:
a

[1, 2, 3]

In [122]:
a[1]=4

[1, 4, 3]

In [125]:
b

[1, 4, 3]

In [126]:
num1=10
num1

10

In [127]:
num2=num1
num1, num2

(10, 10)

In [128]:
num1=11
num1, num2

(11, 10)

In [129]:
a = [1, 2, 3]
b = a[:]
a[1] = 4
a

[1, 4, 3]

In [130]:
b

[1, 2, 3]

## copy

In [133]:
from copy import copy
a=[1,2,3]
a

[1, 2, 3]

In [136]:
b = copy(a)
id(a), id(b)

(2753567310272, 2753594982592)

In [137]:
b is a

False

## 변수를 만드는 여러가지 방법

In [138]:
a, b = ('python', 'life')

In [139]:
a

'python'

In [140]:
b

'life'

In [141]:
(a, b) = 'python', 'life'

In [149]:
a,b

('python', 'python')

In [144]:
[a, b] = ['python', 'life']

In [147]:
a,b

('python', 'life')

In [148]:
 a = b = 'python'